File: C#/Parser.cs
namespace MonadParser
{
    public class Parser<T>
    {
        private readonly Func<string, Maybe<(T, string)>> fun;

        public Parser(Func<string, Maybe<(T, string)>> fun)
        {
            this.fun = fun;
        }

        public Maybe<(T, string)> Parse(string inp) => fun(inp);

        // Functor
        public Parser<TResult> Map<TResult>(Func<T, TResult> mapping)
        {
            Parser<T> self = this;
            return new Parser<TResult>(inp => self.Parse(inp).Map(pair => (mapping(pair.Item1), pair.Item2)));
        }

        // Monad
        public Parser<TResult> FlatMap<TResult>(Func<T, Parser<TResult>> fp)
        {
            Parser<T> self = this;
            return new Parser<TResult>(inp => self.Parse(inp).FlatMap(pair => fp(pair.Item1).Parse(pair.Item2)));
        }

        public Parser<U> Skip<U>(Parser<U> p) => FlatMap(_ => p);

        public Parser<U> Skip<U>(Func<Parser<U>> fp) => FlatMap(_ => fp());

        // Applicative
        public static Parser<T> Pure(T value) => new Parser<T>(inp => Maybe.Some((value, inp)));

        public static Parser<TResult> Apply<TResult>(Parser<Func<T, TResult>> pf, Func<Parser<T>> fq) => pf.FlatMap(func => fq().Map(func));

        // Alternative
        public static Parser<T> Empty() => new Parser<T>(inp => Maybe<(T, string)>.Empty());

        public Parser<T> OrElse(Parser<T> p)
        {
            Parser<T> self = this;
            return new Parser<T>(inp => self.Parse(inp) | (() => p.Parse(inp)));
        }

        public static Parser<T> operator |(Parser<T> p, Parser<T> other) => p.OrElse(other);

        public static Parser<string> Optional(Parser<string> p) => p | Parser<string>.Pure("");

        public static Parser<string> Optional(Parser<char> p) => Optional(p.Map(c => c.ToString()));

        public Parser<string> Some() => Parser<string>.Apply(Map<Func<string, string>>(c => s => c + s), Many);

        public Parser<string> Many() => Optional(Some());

        public static readonly Parser<char> anyChar = new Parser<char>(inp =>
            inp.Length > 0 ?
                Maybe<(char, string)>.Pure((inp[0], inp.Substring(1))) :
                Maybe<(char, string)>.Empty());

        public static Parser<char> Satisfy(Func<char, bool> pred) => anyChar.FlatMap(c => pred(c) ? Parser<char>.Pure(c) : Parser<char>.Empty());

        public static Parser<string> spaces => Satisfy(Char.IsWhiteSpace).Many();

        public static Parser<T> Between<Open, Close>(Parser<Open> open, Parser<Close> close, Parser<T> p) =>
            open.Skip(p).FlatMap(e => close.Skip(Parser<T>.Pure(e)));

        public static Parser<T> Between<Open, Close>(Parser<Open> open, Parser<Close> close, Func<Parser<T>> fp) =>
            open.Skip(fp).FlatMap(e => close.Skip(Parser<T>.Pure(e)));

        public Parser<T> Token() => Between(spaces, spaces, this);

        private static Parser<T> Rest(Func<Parser<T>> fval, Func<T, Parser<T>> ff, Parser<Func<T, T, T>> op, T a) =>
            op.FlatMap(f => fval().FlatMap(b => ff(f(a, b)))) | Parser<T>.Pure(a);

        public Parser<T> Rest_l(Parser<Func<T, T, T>> op, T a)
        {
            Parser<T> self = this;
            return Rest(() => self, b => self.Rest_l(op, b), op, a);
        }

        public Parser<T> Rest_r(Parser<Func<T, T, T>> op, T a)
        {
            Parser<T> self = this;
            return Rest(() => self.Chainr1(op), Parser<T>.Pure, op, a);
        }

        public Parser<T> Chainl1(Parser<Func<T, T, T>> op) => FlatMap(a => Rest_l(op, a));

        public Parser<T> Chainr1(Parser<Func<T, T, T>> op) => FlatMap(a => Rest_r(op, a));
    }
}

File: julia/Parser.jl
module MonadParser

using Monads, Parsers
import Monads: mbind, (>>)

import Base: *, /, |, empty, skip

export Parser, expr, my_parse, mreturn, and_then, satisfy, optional_s, optional_c, between, token, some, many, chainl1, chainr1

# Correct an error in the Monads module
mbind(f::Function, m::Maybe)::Maybe = isa(m.value, Nothing) ? Maybe(nothing) : f(m.value)

and_then(m, f::Function) = mbind(f, m)

(/)(f::Function, m::Monad) = fmap(f, m)

(*)(mf::M, m::Function) where {M<:Monad} = and_then(mf, f -> f / m())
#@mdo M begin
#    f <- mf
#    f / m()
#end

# Alternaive for Maybe
empty(::Type{Maybe})::Maybe = Maybe(nothing)

(|)(m::Maybe, f::Function)::Maybe = !isa(m.value, Nothing) ? m : f()

## Parser
struct Parser <: Monad
    unp :: Function # String -> Maybe{(T, String)}
end

my_parse(p::Parser, inp::String)::Maybe = p.unp(inp)

# Monad
mreturn(Parser, val)::Parser = Parser(inp::String -> Maybe((val, inp)))

# p >>= f = P $ \inp -> do (x, out) <- my_parse p inp; my_parse (f x) out
mbind(f::Function, p::Parser)::Parser = Parser(inp ->
    and_then(my_parse(p, inp), pair -> my_parse(f(first(pair)), last(pair))))

fmap(f::Function, p::Parser)::Parser = and_then(p, x -> mreturn(Parser, f(x)))

# Alternaive for Parser
# empty = P $ \_ -> Nothing
empty(::Type{Parser})::Parser = Parser(_ -> Maybe(nothing))

# p <|> q = P $ \inp -> my_parse p inp <|> my_parse q inp
(|)(p::Parser, q::Parser)::Parser = Parser(inp -> my_parse(p, inp) | () -> my_parse(q, inp))

anyChar = Parser(inp -> length(inp) == 0 ? Maybe(nothing) : Maybe((inp[1], inp[2:length(inp)])))

satisfy(f::Function)::Parser = and_then(anyChar, c -> f(c) ? mreturn(Parser, c) : empty(Parser))

optional_s(p::Parser)::Parser = p | mreturn(Parser, "")

optional_c(p::Parser)::Parser = optional_s((c -> string(c)) / p)

some(p::Parser)::Parser = (c -> s -> c * s) / p * () -> many(p)

many(p::Parser)::Parser = optional_s(some(p))

# spaces = many $ satisfy isSpace
spaces = many(satisfy(c::Char -> isspace(c)))

skip(p::Parser, fq::Function)::Parser = and_then(p, _ -> fq())

between(open::Parser, close::Parser, fp::Function)::Parser =
    and_then(skip(open, fp), x -> skip(close, () -> mreturn(Parser, x)))

token(p::Parser)::Parser = between(spaces, spaces, () -> p)

# rest :: Parser a -> (a -> Parser a) -> Parser (a -> a -> a) -> a -> Parser a
rest(p::Function, ff::Function, op::Parser, a)::Parser =
    and_then(op, f::Function -> and_then(p(), b -> ff(f(a, b)))) | mreturn(Parser, a)

# chainl1 :: Parser a -> Parser (a -> a -> a) -> Parser a
function chainl1(p::Parser, op::Parser)::Parser
    # rest_l a = rest p rest_l op a
    rest_l(a) = rest(() -> p, rest_l, op, a)

    and_then(p, a -> rest_l(a))
end

# chainr1 :: Parser a -> Parser (a -> a -> a) -> Parser a
function chainr1(p::Parser, op::Parser)::Parser
    # scan = do { a <- p; rest_r a }
    scan() = and_then(p, a -> rest_r(a))

    # rest_r a = rest scan return op a
    rest_r(a) = rest(scan, b -> mreturn(Parser, b), op, a)

    scan()
end

end # module MonadParser
